arch/xtensa/src/esp32s3/esp32s3_spi.c

```diff
@@ -0,0 +1,1219 @@
+ *          number of bits selected for the SPI interface.
+ *
+ * Returned Value:
+ *   Received value.
+ *
+ ****************************************************************************/
+
+static uint32_t esp32s3_spi_poll_send(struct esp32s3_spi_priv_s *priv,
+                                      uint32_t wd)
+{
+  uint32_t val;
+
+  putreg32((priv->nbits - 1), SPI_MS_DLEN_REG(priv->config->id));
+
+  putreg32(wd, SPI_W0_REG(priv->config->id));
+
+  /* Trigger start of user-defined transaction for master. */
+
+  esp32s3_spi_set_regbits(SPI_CMD_REG(priv->config->id), SPI_UPDATE_M);
+
+  while ((getreg32(SPI_CMD_REG(priv->config->id)) & SPI_UPDATE_M) != 0)
+    {
+      ;
+    }
+
+  esp32s3_spi_set_regbits(SPI_CMD_REG(priv->config->id), SPI_USR_M);
+
+  /* Wait for the user-defined transaction to finish. */
+
+  while ((getreg32(SPI_CMD_REG(priv->config->id)) & SPI_USR_M) != 0)
+    {
+      ;
+    }
+
+  val = getreg32(SPI_W0_REG(priv->config->id));
+
+  spiinfo("send=0x%" PRIx32 " and recv=0x%" PRIx32 "\n", wd, val);
+
+  return val;
+}
+
+/****************************************************************************
+ * Name: esp32s3_spi_send
+ *
+ * Description:
+ *   Send one word on SPI.
+ *
+ * Input Parameters:
+ *   dev - Device-specific state data
+ *   wd  - The word to send. The size of the data is determined by the
+ *         number of bits selected for the SPI interface.
+ *
+ * Returned Value:
+ *   Received value.
+ *
+ ****************************************************************************/
+
+static uint32_t esp32s3_spi_send(struct spi_dev_s *dev, uint32_t wd)
+{
+  struct esp32s3_spi_priv_s *priv = (struct esp32s3_spi_priv_s *)dev;
+
+  return esp32s3_spi_poll_send(priv, wd);
+}
+
+/****************************************************************************
+ * Name: esp32s3_spi_poll_exchange
+ *
+ * Description:
+ *   Exchange a block of data from SPI.
+ *
+ * Input Parameters:
+ *   priv     - SPI private state data
+ *   txbuffer - A pointer to the buffer of data to be sent
+ *   rxbuffer - A pointer to the buffer in which to receive data
+ *   nwords   - The length of data that to be exchanged in units of words.
+ *              The wordsize is determined by the number of bits-per-word
+ *              selected for the SPI interface. If nbits <= 8, the data is
+ *              packed into uint8_t's; if nbits >8, the data is packed into
+ *              uint16_t's
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void esp32s3_spi_poll_exchange(struct esp32s3_spi_priv_s *priv,
+                                      const void *txbuffer,
+                                      void *rxbuffer,
+                                      size_t nwords)
+{
+  const uint32_t total_bytes = nwords * (priv->nbits / 8);
+  uintptr_t bytes_remaining = total_bytes;
+  uint8_t *tp = (uint8_t *)txbuffer;
```

Comment:
can we still keep `const`?
```suggestion
  const uint8_t *tp = (const uint8_t *)txbuffer;
```